scanner/string.go

```diff
@@ -0,0 +1,308 @@
+		s.next()
+		if s.char != '-' {
+			s.failf("invalid escape")
+			return skip
+		}
+		skip++
+		s.next()
+		c = decodeCtrlEsc(s.char)
+	case 'c':
+		skip++
+		s.next()
+		c = decodeCtrlEsc(s.char)
+	}
+	for i := 1; i < n; i++ {
+		skip++
+		s.next()
+	}
+	replace(s, c, skip)
+	return skip
+}
+
+func decodeCtrlEsc(c byte) byte {
+	if c == '?' {
+		return 0x7f
+	}
+	if v := escapes[c]; v != 0 {
+		c = v
+	}
+	return c & 0x9f
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func decodeOctalEsc(s *Scanner) (n int, v byte) {
+	m := min(3, len(s.src)-s.offset)
+	for n = 0; n < m; n++ {
+		c := s.src[s.offset+n]
+		if !token.IsOctadecimal(c) {
+			break
+		}
+		v = v*8 + (c - '0')
+	}
+	if n == 0 {
+		s.failf("invalid octal escape")
+	}
+	return
+}
+
+func decodeHexEsc(s *Scanner) (n int, v byte) {
+	m := min(2, len(s.src)-s.offset)
+	for n = 0; n < m; n++ {
+		c := s.src[s.offset+n]
+		var d byte
+		switch {
+		case '0' <= c && c <= '9':
+			d = c - '0'
+		case 'A' <= c && c <= 'F':
+			d = c - 'A' + 10
+		case 'a' <= c && c <= 'f':
+			d = c - 'a' + 10
+		default:
+			if n == 0 {
+				s.failf("invalid hex escape")
+			}
+			return
+		}
+		v = v*16 + d
+	}
+	return
+}
+
+func stateInsertStmts(s *Scanner) (pos int, t token.Token, literal []byte) {
+	s.begin = s.offset
+	if s.char == '}' {
+		s.next()
+		s.popCtx()
+		return s.begin, token.InsertEnd, nil
+	}
+	return stateCompStmts(s)
+}
+
+func scanSingleQuotedString(s *Scanner, term byte, head int) (token.Token, []byte) {
+	var skip int
+	for s.char != term && s.err == nil {
+		if s.char == '\\' {
+			s.next()
+			if s.char == '\\' || s.char == term {
+				skip++
+			}
+		}
+		replace(s, s.char, skip)
+	}
+	return token.String, s.src[s.begin+head : s.offset-skip]
+}
+
+func scanHeredocBegin(s *Scanner) (token.Token, []byte) {
+	indent := false
+	termBegin := s.offset
+	switch c := s.char; {
+	case token.IsLetter(c) || c == '_':
+		s.next()
+	case token.IsDecimal(c):
+		if s.ctx.nospace {
+			return token.Continue, nil
+		}
+		s.next()
+	case c == '-':
+		if s.ctx.nospace {
+			return token.Continue, nil
+		}
+		indent = true
+		s.next()
+		termBegin = s.offset
+	default:
+		return token.Continue, nil
+	}
+
+	var quote byte
+	if s.char == '\'' {
+		quote = s.char
+		s.next()
+		termBegin = s.offset
+	}
+	for token.IsIdent(s.char) {
+		s.next()
+	}
+	term := s.src[termBegin:s.offset]
+	if quote != 0 {
+		if s.char != quote {
+			s.failf("invalid heredoc identifier")
+		}
+		s.next()
+	}
+	s.pushCtx(stateHeredocFirstLine(term, indent))
+	return token.HeredocBegin, s.src[s.begin:s.offset]
+}
+
+func stateHeredocFirstLine(term []byte, indent bool) stateScanFunc {
+	return func(s *Scanner) (int, token.Token, []byte) {
+		if s.char == '\n' {
+			begin := s.offset
+			s.next()
+			s.popCtx()
+			s.pushCtx(stateInHeredoc(term, indent))
+			return begin, token.NewLine, nil
+		}
+		return stateCompStmts(s)
+	}
+}
+
+func stateInHeredoc(term []byte, indent bool) stateScanFunc {
```

Comment:
__[gometalinter]__
* [gocyclo] cyclomatic complexity 14 of function stateInHeredoc() is high (> 10)

<!-- Comment by [SideCI](https://sideci.com) -->